x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsDbSourceModeMigrationIT.java

```diff
@@ -0,0 +1,286 @@
+                  "type": "date",
+                  "format": "epoch_millis"
+                },
+                "message": {
+                  "type": "text"
+                },
+                "method": {
+                  "type": "keyword"
+                },
+                "hits": {
+                  "type": "long"
+                }
+              }
+            }
+          }
+        }""";
+
+    public static final String STORED_SOURCE_COMPONENT_TEMPLATE = """
+        {
+          "template": {
+            "settings": {
+              "index": {
+                "mapping.source.mode": "stored"
+              }
+            }
+          }
+        }""";
+
+    public static final String SYNTHETIC_SOURCE_COMPONENT_TEMPLATE = """
+        {
+          "template": {
+            "settings": {
+              "index": {
+                "mapping.source.mode": "synthetic"
+              }
+            }
+          }
+        }""";
+
+    @ClassRule()
+    public static ElasticsearchCluster cluster = ElasticsearchCluster.local()
+        .distribution(DistributionType.DEFAULT)
+        .module("constant-keyword")
+        .module("data-streams")
+        .module("mapper-extras")
+        .module("x-pack-aggregate-metric")
+        .module("x-pack-stack")
+        .setting("xpack.security.enabled", "false")
+        .setting("xpack.otel_data.registry.enabled", "false")
+        .setting("xpack.license.self_generated.type", "trial")
+        .setting("cluster.logsdb.enabled", "true")
+        .build();
+
+    @Override
+    protected String getTestRestCluster() {
+        return cluster.getHttpAddresses();
+    }
+
+    @Before
+    public void setup() throws Exception {
+        client = client();
+        waitForLogs(client);
+    }
+
+    private RestClient client;
+
+    public void testSwitchFromStoredToSyntheticSource() throws IOException {
+        assertOK(putComponentTemplate(client, "my-logs-mapping", MAPPING_COMPONENT_TEMPLATE));
+        assertOK(putComponentTemplate(client, "my-logs-original-source", STORED_SOURCE_COMPONENT_TEMPLATE));
+
+        assertOK(putTemplate(client, "my-logs", INDEX_TEMPLATE));
+        assertOK(createDataStream(client, "my-logs-ds-test"));
+
+        var initialSourceMode = (String) getSetting(
+            client,
+            getDataStreamBackingIndex(client, "my-logs-ds-test", 0),
+            "index.mapping.source.mode"
+        );
+        assertThat(initialSourceMode, equalTo("stored"));
+        var initialIndexMode = (String) getSetting(client, getDataStreamBackingIndex(client, "my-logs-ds-test", 0), "index.mode");
+        assertThat(initialIndexMode, equalTo("logsdb"));
+
+        var indexedWithStoredSource = new ArrayList<XContentBuilder>();
+        var indexedWithSyntheticSource = new ArrayList<XContentBuilder>();
+        for (int i = 0; i < 10; i++) {
+            indexedWithStoredSource.add(generateDoc());
+            indexedWithSyntheticSource.add(generateDoc());
```

Comment:
Shouldn't we be adding the same doc in both?

Reply:
We will see the union of these two lists in search so we want them to be different to assert nothing gets lost.